python/tests/test_minelogy.py

```diff
@@ -111,11 +111,11 @@ def test_minelogy(self):
             )
             ''')
         output = utils.interpret('(how-get cobblestone)')
-        self.assertEqual(repr(output[0]),
-            '(do-mine ((: stone type) (: stone variant)))')
+        self.assertEqual(repr(output),
+            '[(let $t (get-ingredients cobblestone) (do-craft $t)), (do-mine ((: stone type) (: stone variant)))]')
```

Comment:
I wonder, why `(let $t (get-ingredients cobblestone) (do-craft $t))` remains here. Was it there before, but as `output[1]`, which was ignored? `(get-ingredients cobblestone)` should return no result as it calls `match`. There is a general problem, as we discussed, to understand if "no result" should be propagated to the parent expression, so it also returns empty set of results, and when it means that the expression just should not be reduced. It might be possible to control this behaviour by the return type (or type annotation?) of the functional atom (in this case, `let`). Maybe, it worths somehow marking this example for further consideration within this general question.

Reply:
I investigated it further and found that I changed behavior while unified alternative planning for the argument reduction and matching/execution. The `(get-ingredients cobblestone)` is reduced to empty result because it is transformed to the grounded `match` operation and it in turn returns empty result. There is no `Err` which is returned by embedded matching. I reverted behavior to the previous state in #99.

Reply:
Nice. But in any case, it seems that we still need to sort out, when we do expect an empty result instead of a not reduced expression besides direct calls to `match`. For example, for
```
(= (if! True $x) $x)
```
we may want to return no result if the first argument is `False`. Now, `println! (if! False (+ 1 1)))` will print `(if! False (+ 1 1))`. There can be different approaches to it:
- to have real `void` or `abort` ("element" of `Void`). This "contradicts" to type and category theory, which say that no function can return "no result" (constructing a function of type `A -> Void` proves that `A` is false)... but who cares... so, `(= (if! False $_) void)` will explicitly abort this branch of evaluation of the parent expression... oh, well, we already have `nop` for this, which differs from `void` only in that it is a function..
- wrap the expression into a specific function forcing the interpreter to find equality for the wrapped expression or abort, e.g. `(force-reduce (if! False (+ 1 1)))`. The advantage is that we can select ad hoc behavior.
- explicitly define behavior via types (this might also make implementing `force-reduce` naturally).